decisions/0001-bundle-granularity.rst

```diff
@@ -0,0 +1,199 @@
+==================
+
+Data about the OLX contents of a Bundle (e.g. tagging, search indexing) would
+need to be associated to the tuple of (BundleVersion, XBlock Directory). Asset
+association also happens at this level of granularity.
+
+Publishing Transactions
+=======================
+
+When updating a course, a new Snapshot would be created by a Draft, with all the
+changes for any given set of XBlocks. Once the Snapshot successfully completes,
+a new BundleVersion of the Course would be created to point to that Snapshot. An
+interruption in the publishing process during Snapshot creation could result in
+an orphaned Snapshot that's no pointed to by a BundleVersion, but shouldn't
+result in a broken or inconsitent state.
+
+Hierarchy Representation
+========================
+
+The list of XBlock content is represented as a flat set of directories, and all
+navigational hierarchy has to be interpreted by OLX-aware clients. A student's
+path through a course is an LMS/Compositor level concern that structural OLX is
+an input to, and the storage layer of Blockstore shouldn't need to model it.
+
+Reusing Containers
+==================
+
+To re-use a Unit, you would first make a Link to the Bundle where that Unit came
+from, and then specify the Link prefix before the directory where the Unit comes
+from::
+
+    <xblock-include definition="arch_hangout_videos/video+blockstore_overview"/>
+
+This allows for arbitrary reuse at varying levels of granularity.
+
+Intentionality and Tracking Re-use
+==================================
+
+A drawback of approach is that it requires more explicit intentionality in
+course design in order for its BundleVersion dependency tracking to be
+meaningful. You could borrow a single leaf block or container block from another
+Course, but Blockstore itself would only know that the link between the
+BundleVersions existed, not the specific items that were used. This problem is
+significantly lessened in the case of Content Libraries, since each problem has
+its own Bundle there.
+
+Another way to look at it is that Blockstore's tracking of reuse is for
+update notifications, dependency checking, and licensing enforcement. Finer
+granularity measurements are better done by something more OLX-aware. For
+instance, if I'm making a CCX that uses a sequence but I hide a Unit that isn't
+relevant to my students, should that count towards my reuse of that particular
+unit? If someone Links to an entire Unit, but really they just wanted to make
+a reference to one image in one leaf of that Unit, is it Blockstore's job to
+understand the references well enough to understand that?
+
+If we start with this kind of tracking being at the core of Blockstore's data
+model, then OLX awareness and coupling will slowly work its way into the system.
+
+Performance
+===========
+
+Storing containers in a granular way makes certain kinds of concurrent
+operations simpler, like Drafts independently publishing Units and Chapters.
```

Comment:
Unless we have identified a need for independent re-use of Sections and Subsections, I'd love to just roll that up into a single `outline.xml` file (which can be transformed via a diff or a JSON ruleset in each CCX course bundle [`ADD my-custom-section AFTER course-intro`, `REMOVE advanced-topic-section`] )

Reply:
I think I'm more hung up on the problem from a publishing point of view. Say you want to publish a chapter, and only one chapter. If the data describing that chapter is wrapped up with the rest of the outline, then you have to be able to break that up when you store it. Which is possible, but seems like you could run into weird concurrency edge cases.

I'm totally +1 on having some separate ruleset to express transforms to structure. But again, I'd like the parsing knowledge of that to live outside of Blockstore. I can easily see things like "add a node here" evolve to things like, "get me this, but remove all the survey blocks".

Reply:
We could stick to the current convention that changes to the course outline are immediate, but changes within Units or smaller can have draft-publish workflow. However, it would be nice to do better.

This is probably too complex/messy to bother with for now, but it could offer a potential solution in the future: Imagine if in order to support the CCX use case, we expect that master course bundles have an `outline.xml` and each CCX course bundle links to the master bundle and also has its `outline-changes.xml` which has the list of customizations unique to that CCX course (add/remove/re-order sections via rules). If we've already implemented all that logic, then we could use the same thing for draft-publish: In the published version of the course would sit the normal `outline.xml` and in each draft version of the course bundle (there could be many), would be an unmodified `outline.xml` plus an `outline-changes.xml` with the changes specific to that draft version of the course. Then, we get to re-use code, we don't add any functionality to Blockstore to support that use case, and we get section/subsection-level publishing.

You touch on something that's a point of concern though whether the outline data is in one file or many: what if two authors concurrently change a section by adding/removing different subsections? S3 supports atomic writes and read-after-write consistency, but it doesn't support locking, so there's no way to read-update-write while blocking other reads. Would we want to implement a basic form of write locking at the Blockstore level?

Reply:
> We could stick to the current convention that changes to the course outline are immediate, but changes within Units or smaller can have draft-publish workflow. However, it would be nice to do better.

I really hate that we do this. As far as I can tell, the reason this exists today is because Old Mongo had technical limitations that made it difficult to avoid and then Split came along and build a layer to emulate the brokenness to be compatible.

> <snip> course outline transform idea.

Unifying CCX and draft transformation of the outline is a really interesting idea. Could you expand on the details in #26?

> You touch on something that's a point of concern though whether the outline data is in one file or many: what if two authors concurrently change a section by adding/removing different subsections? S3 supports atomic writes and read-after-write consistency, but it doesn't support locking, so there's no way to read-update-write while blocking other writes. Would we want to implement a basic form of write locking at the Blockstore level?

I'd rather avoid creating lock files if at all possible. I think that the number of concurrent authors are small enough where it's realistic to expect them to coordinate with each other, if we can do something even as dumb as exposing an activity log so they know someone else is working on it. Higher concurrency scenarios, we punt to the git course development workflow.

But we should definitely be explicit and map out the failure modes. I'll work on that this evening.

Reply:
> I'd rather avoid creating lock files if at all possible.

I definitely wouldn't suggest lock files. What I had in mind was obtaining a bundle-specific [MySQL advisory lock](https://dev.mysql.com/doc/refman/5.7/en/locking-functions.html) before certain types of bundle write operations, to prevent race conditions.

The main thing I was worried about is:
* Client 1 deletes `file1.olx`
* Client 2 adds `file2.olx`
* Blockstore session for client 1 opens `bundle.json` for reading
* Blockstore session for client 2 opens `bundle.json` for reading
* Blockstore session for client 1 overwrites `bundle.json` without `file1.olx`
* Blockstore session for client 2 overwrites `bundle.json` with `file2.olx` (and with `file1.olx`)

Now `bundle.json` contains a reference to `file1.olx` but it no longer exists on S3.